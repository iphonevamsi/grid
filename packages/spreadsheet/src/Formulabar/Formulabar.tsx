import React, { memo, forwardRef, useImperativeHandle, useRef } from "react";
import {
  InputGroup,
  InputLeftAddon,
  Input,
  useColorMode,
  useTheme,
  Box,
} from "@chakra-ui/core";
import {
  DARK_MODE_COLOR,
  FORMULABAR_LEFT_CORNER_WIDTH,
  FORMULA_FONT,
  SYSTEM_FONT,
  isAFormula,
  FORMULA_FONT_SIZE,
  DEFAULT_FONT_SIZE,
  pointToPixel,
  DEFAULT_FORMULABAR_HEIGHT,
  DEFAULT_FONT_COLOR,
} from "./../constants";
import Resizer from "../Resizer";
import TextEditor, { EditableRef } from "./../Editor/TextEditor";
import { FormulaChangeProps } from "../Spreadsheet";
import { Direction } from "@rowsncolumns/grid";

interface FormulabarProps {
  onChange?: (value: string) => void;
  onKeyDown?: (
    e: React.KeyboardEvent<HTMLInputElement | HTMLDivElement>
  ) => void;
  onFocus?: (e: React.FocusEvent<HTMLInputElement | HTMLDivElement>) => void;
  onBlur?: (e: React.FocusEvent<HTMLInputElement | HTMLDivElement>) => void;
  onSubmit?: (value: string, direction?: Direction) => void;
  onCancel?: () => void;
  value?: string;
  isFormulaMode?: boolean;
  locked?: boolean;
  height?: number;
  onChangeHeight?: (value: number) => void;
  supportedFormulas?: string[];
  onFormulaChange?: (props: FormulaChangeProps) => void;
  disabled?: boolean;
}

export type RefAttribute = {
  ref?: React.Ref<EditableRef | null>;
};

const Formulabar: React.FC<FormulabarProps & RefAttribute> = memo(
  forwardRef((props, forwardedRef) => {
    const {
      value = "",
      onChange,
      onKeyDown,
      onFocus,
      onBlur,
      isFormulaMode = false,
      locked,
      height = DEFAULT_FORMULABAR_HEIGHT,
      onChangeHeight,
      supportedFormulas,
      onFormulaChange,
      onSubmit,
      onCancel,
      disabled,
    } = props;
    const inputRef = useRef<EditableRef>(null);
    const isFormula = isAFormula(value) || isFormulaMode;
    const { colorMode } = useColorMode();
    const theme = useTheme();
    const isLightMode = colorMode === "light";
    const backgroundColor = isLightMode ? "white" : DARK_MODE_COLOR;
    const color = isLightMode ? DARK_MODE_COLOR : "white";
    const borderColor = isLightMode
      ? theme.colors.gray[300]
      : theme.colors.gray[600];

    return (
      <Box position="relative">
        <InputGroup
          size="sm"
          borderTopWidth={1}
          borderTopStyle="solid"
          borderTopColor={borderColor}
          height={`${height}px`}
        >
          <InputLeftAddon
            width={FORMULABAR_LEFT_CORNER_WIDTH}
            justifyContent="center"
            bg={backgroundColor}
            color={color}
            fontSize={12}
            fontStyle="italic"
            borderTopWidth={0}
            borderBottomWidth={0}
            borderRightWidth={1}
            size="sm"
            borderRadius={0}
            children="fx"
            height="auto"
            userSelect="none"
            borderLeftColor={borderColor}
            borderRightColor={borderColor}
          />
          <div
            style={{
              display: "flex",
              flex: 1,
              position: "relative",
              marginTop: 4,
              marginLeft: 6,
              paddingLeft: 1,
            }}
          >
            <TextEditor
              ref={forwardedRef}
              isFormulaMode={isFormula}
              value={value}
              fontFamily={isFormula ? FORMULA_FONT : SYSTEM_FONT}
              fontSize={pointToPixel(DEFAULT_FONT_SIZE) as number}
              onChange={onChange}
              onFocus={onFocus}
              onBlur={onBlur}
              autoFocus={false}
              onSubmit={onSubmit}
              onCancel={onCancel}
              tearDown
              scale={1}
              color={DEFAULT_FONT_COLOR}
              wrapping
              horizontalAlign="left"
              disabled={locked || disabled}
              supportedFormulas={supportedFormulas}
              lineHeight="13px"
              overflow="auto"
              onFormulaChange={onFormulaChange}
            />
          </div>
        </InputGroup>

        <Resizer
          minTop={DEFAULT_FORMULABAR_HEIGHT}
          top={DEFAULT_FORMULABAR_HEIGHT}
          onDrag={onChangeHeight}
        />
      </Box>
    );
  })
);

export default Formulabar;
