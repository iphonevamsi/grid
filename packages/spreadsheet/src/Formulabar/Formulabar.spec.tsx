import React, { useState, useRef } from "react";
import { domRenderer, cleanup, fireEvent, act } from "./../utils/test-utils";
import Formulabar from "./Formulabar";

global.document.execCommand = jest.fn();

describe("FormulaBar", () => {
  afterEach(cleanup);
  it("renders spreadsheet", () => {
    const renderGrid = () => domRenderer(<Formulabar />);
    expect(renderGrid).not.toThrow();
  });
  it("matches snapshot", () => {
    const { asFragment } = domRenderer(<Formulabar />);
    expect(asFragment()).toMatchSnapshot();
  });

  // it("can trigger onchange", () => {
  //   const onChange = jest.fn();
  //   const { getByLabelText } = domRenderer(<Formulabar onChange={onChange} />);
  //   const input = getByLabelText("value-input") as HTMLInputElement;
  //   fireEvent.change(input, { target: { value: "hello" } });
  //   expect(onChange).toBeCalled();
  //   expect(onChange).toBeCalledWith("hello");
  // });
});
