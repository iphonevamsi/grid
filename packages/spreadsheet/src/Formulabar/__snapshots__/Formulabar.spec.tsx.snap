// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormulaBar matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-79elbk"
  >
    <div
      class="css-f7ejuk"
    >
      <div
        class="css-19p1yte"
      >
        fx
      </div>
      <div
        style="display: flex; flex: 1; position: relative; margin-top: 4px; margin-left: 6px; padding-left: 1px;"
      >
        <div
          style="font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji; font-size: 13px; font-weight: normal; width: 100%; height: 100%; padding: 0px 1px; margin: 0px; box-sizing: border-box; border-width: 0px; outline: none; resize: none; overflow: auto; vertical-align: top; background: transparent; color: rgb(0, 0, 0); white-space: pre-wrap; text-align: left; line-height: 13px; text-decoration: none; cursor: text; flex: 1; display: flex;"
        >
          <div
            contenteditable="true"
            data-gramm="false"
            data-slate-editor="true"
            data-slate-node="value"
            role="textbox"
            style="outline: none; white-space: pre-wrap; word-wrap: break-word; flex: 1;"
          >
            <div
              data-slate-node="element"
              style="position: relative;"
            >
              <span
                data-slate-node="text"
              >
                <span
                  data-slate-leaf="true"
                >
                  <span
                    data-slate-length="0"
                    data-slate-zero-width="n"
                  >
                    ï»¿
                    <br />
                  </span>
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      style="user-select: none; height: 4px; position: absolute; left: 0px; margin-top: -4px; top: 24px; right: 0px; z-index: 1; cursor: ns-resize;"
    />
  </div>
</DocumentFragment>
`;
